04-evm-mapping.md

```diff
@@ -0,0 +1,307 @@
+
+**Class 2 and 4 address derivation**
+
+Class 2 and 4 addresses are protocol-derived, by hashing the relevant cryptographic or input material.
+
+With a high probability, the hash function of class 4 will remain BLAKE2, with a 256-bit digest size. Class 2 will continue relying on blake2b-160.
+
+1. For secp256k1 account actors (class 1), the preimage is the pubkey. The value of the class 2 address is identical to that of class 1.
+2. For BLS account actors (class 3), the preimage is the pubkey.
+3. For FVM native actors, the preimage is `sender || nonce || # of actors created during message execution`.
+4. For EVM foreign actors, the preimage is inherited from CREATE and CREATE2.
+
+Note that preimages are not user-controlled, but some constituents of them may be (e.g. EVM CREATE2 opcode).
+
+**Resulting address space taxonomy**
+
+| Class | Desc                              | Actor type | Payload width | Total width | Payload value                                         | Usage                                                                                | Stable? |
+| ----- | --------------------------------- | ---------- | ------------- | ----------- | ----------------------------------------------------- | ------------------------------------------------------------------------------------ | ------- |
+| 0     | ID address                        | All        | 1-9 bytes     | 2-10 bytes  | uvarint64 counter                                     | Internal, compact representation in state tree; unsafe to use externally until final | N       |
+| 1     | secp256k1 pubkey (account actor)  | Account    | 20 bytes      | 21 bytes    | blake2b-160 hash of secp256k1 pubkey                  | Externally, to refer to an account actor with its pubkey                             | Y       |
+| 2     | aliased universal actor address   | All        | 20 bytes      | 21 bytes    | protocol-derived from relevant cryptographic material | Externally and internally to refer to any actor                                      | Y       |
+| 3     | bls pubkey (account actor)        | Account    | 48 bytes      | 49 bytes    | inlined bls public key                                | Externally, to refer to an account actor with its pubkey                             | Y       |
+| 4     | canonical universal actor address | All        | 32 bytes      | 33 bytes    | protocol-derived from relevant cryptographic material | Externally and internally to refer to any actor                                      | Y       |
+
+## Gas accounting and execution halt semantics
+
+The execution halt is determined by Filecoin gas and not by EVM gas. Therefore, EVM runtime is made to run with unlimited gas. The FVM is responsible for metering execution and halting it when gas limits are exceeded. Refer to the [Gas accounting](01-architecture.md#gas-accounting) section of the Architecture doc for more details.
```

Comment:
This isn't quite what we discussed. Top-level messages will be run with infinite (well, max int) EVM gas. But an EVM contract may call other contracts with some limited amount of gas and _that_ gas limit should be respected.

Basically:

1. We obey both gas models at the same time.
2. The EVM gas model starts with infinite gas.

I believe this may be important for contracts that need to be able to invoke untrusted contracts with some limited amount of gas.

----

Aside: I'm wondering if we can optimize EVM execution by being a bit looser about gas usage. If infinite gas is available, we could go down a codepath that doesn't bother tracking gas.

Reply:
Alternatively, we could let the message sender specify the amount of EVM gas. Basically, when doing gas estimation, the caller would try the message with infinite EVM gas. Then, the caller would replace the "infinite" EVM gas with the _used_ EVM gas (times some reasonable overestimation multiplier).

Given a high enough overestimation, this reduces the chances of spuriously running out of EVM gas. However, having a somewhat realistic gas value lets the contract make decisions based on the amount of gas left.

As long as we allow the message sender to specify the EVM gas limit, they can pick their model based on their use-case. If, e.g., they have a contract that really cares about EVM gas, they can pass in a gas estimation. Otherwise, they can pass in "infinite" gas.